kaldb/src/main/java/com/slack/kaldb/recovery/RecoveryService.java

```diff
@@ -0,0 +1,160 @@
+    this.searchContext = SearchContext.fromConfig(recoveryConfig.getServerConfig());
+
+    // we use a single thread executor to allow operations for this recovery node to queue,
+    // guaranteeing that they are executed in the order they were received
+    this.executorService =
+        Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder().setNameFormat("recovery-service-%d").build());
+
+    Collection<Tag> meterTags = ImmutableList.of(Tag.of("nodeHostname", searchContext.hostname));
+    recoveryNodeReceivedAssignment =
+        meterRegistry.counter(RECOVERY_NODE_RECEIVED_ASSIGNMENT, meterTags);
+    recoveryNodeCompletedAssignment =
+        meterRegistry.counter(RECOVERY_NODE_COMPLETED_ASSIGNMENT, meterTags);
+    recoveryNodeFailedAssignment =
+        meterRegistry.counter(RECOVERY_NODE_FAILED_ASSIGNMENT, meterTags);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting recovery service");
+    metadataStoreService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    recoveryNodeMetadataStore =
+        new RecoveryNodeMetadataStore(metadataStoreService.getMetadataStore(), false);
+    recoveryTaskMetadataStore =
+        new RecoveryTaskMetadataStore(metadataStoreService.getMetadataStore(), false);
+
+    recoveryNodeMetadataStore.createSync(
+        new RecoveryNodeMetadata(
+            searchContext.hostname,
+            Metadata.RecoveryNodeMetadata.RecoveryNodeState.FREE,
+            "",
+            Instant.now().toEpochMilli()));
+    recoveryNodePreviousState = Metadata.RecoveryNodeMetadata.RecoveryNodeState.FREE;
+
+    recoveryNodeListenerMetadataStore =
+        new RecoveryNodeMetadataStore(
+            metadataStoreService.getMetadataStore(), searchContext.hostname, true);
+    recoveryNodeListenerMetadataStore.addListener(recoveryNodeListener());
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Closed recovery service");
+  }
+
+  private KaldbMetadataStoreChangeListener recoveryNodeListener() {
+    return () -> {
+      RecoveryNodeMetadata recoveryNodeMetadata =
+          recoveryNodeMetadataStore.getNodeSync(searchContext.hostname);
+      Metadata.RecoveryNodeMetadata.RecoveryNodeState newRecoveryNodeState =
+          recoveryNodeMetadata.recoveryNodeState;
+
+      if (newRecoveryNodeState.equals(Metadata.RecoveryNodeMetadata.RecoveryNodeState.ASSIGNED)) {
+        LOG.info("Recovery node - ASSIGNED received");
+        recoveryNodeReceivedAssignment.increment();
+        if (!recoveryNodePreviousState.equals(
+            Metadata.RecoveryNodeMetadata.RecoveryNodeState.FREE)) {
+          LOG.warn(
+              "Unexpected state transition from {} to {}",
+              recoveryNodePreviousState,
+              newRecoveryNodeState);
+        }
+        executorService.submit(() -> handleRecoveryTaskAssignment(recoveryNodeMetadata));
+      }
+      recoveryNodePreviousState = newRecoveryNodeState;
```

Comment:
should this be `recoveryNodeCurrentState` ?

Reply:
This is the last known state that we have record of. The next time the `recoveryNodeListener()` is invoked this variable (`recoveryNodePreviousState`) would contain our last observed value which is expected to be different than the current value in ZK (`recoveryNodeMetadata.recoveryNodeState` aka `newRecoveryNodeState`).

This implementation matches that of the `ReadOnlyChunkImpl` and is used to detect unexpected transitions - https://github.com/slackhq/kaldb/blob/master/kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java#L147.

Perhaps a better name for these varianbles would be `fooLastKnownState` or `fooLastObservedState` instead of `fooPreviousState`?

Reply:
I went ahead and renamed these in both places to `lastKnownState` as I think it helps make it more obvious what's going on.